lib/app_profiler/viewer/speedscope_remote_viewer/base_middleware.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module AppProfiler
+  module Viewer
+    class SpeedscopeRemoteViewer < BaseViewer
+      class BaseMiddleware
+        def self.id(file)
+          file.basename.to_s.delete_suffix(".json")
+        end
+
+        def initialize(app)
+          @app = app
+        end
+
+        def call(env)
+          request = Rack::Request.new(env)
+
+          return index(env)                        if request.path_info =~ %r(\A/app_profiler\z)
+          return viewer(env, Regexp.last_match(1)) if request.path_info =~ %r(\A/app_profiler/viewer/(.*)\z)
+          return show(env, Regexp.last_match(1))   if request.path_info =~ %r(\A/app_profiler/(.*)\z)
+
+          @app.call(env)
+        end
+
+        protected
+
+        def id(file)
+          self.class.id(file)
+        end
+
+        def profile_files
+          AppProfiler.profile_root.glob("**/*.json")
+        end
+
+        def render(html)
+          [
+            200,
+            { "Content-Type" => "text/html" },
+            [
+              +<<~HTML,
+                <!doctype html>
+                <html>
+                  <head>
+                    <title>App Profiler</title>
+                  </head>
+                  <body>
+                    #{html}
```

Comment:
I believe rendering like this leaves you vulnerable to XSS if `html` ever had anything malicious in it since ActiveSupport provides the default sanitizations

Reply:
`sanitize` belongs to Action View which IIRC uses https://github.com/rails/rails-html-sanitizer. I think I can use that here.